File: Native/backend/src/utils/databaseValidator.js
import mongoose from 'mongoose';
import { Session } from '../models/Session.js';
import { Driver } from '../models/Driver.js';
import { Vehicle } from '../models/Vehicle.js';
import { Telemetry } from '../models/Telemetry.js';

/**
 * Database Validation Utility
 * Ensures all required collections exist and are writable before allowing operations
 * Prevents silent failures from misconfigured schemas or full databases
 */

/**
 * Validate database connection and collections
 * @returns {Promise<Object>} Validation result with status and errors
 */
export async function validateDatabaseSetup() {
  console.log('🔍 Starting comprehensive database validation...\n');
  
  const errors = [];
  const warnings = [];
  const successful = [];

  try {
    // Step 1: Verify connection
    console.log('1️⃣  Testing MongoDB connection...');
    if (!mongoose.connection || mongoose.connection.readyState === 0) {
      errors.push('❌ MongoDB connection not established');
      return {
        success: false,
        errors,
        warnings,
        successful,
      };
    }
    successful.push('✅ Connected to MongoDB');
    console.log('   ✅ Connected');

    // Step 2: Verify database exists and is accessible
    console.log('\n2️⃣  Verifying database accessibility...');
    
    if (!mongoose.connection.db) {
      errors.push('❌ Database instance not initialized yet');
      console.log('   ❌ Database instance not ready');
      return {
        success: false,
        errors,
        warnings,
        successful,
      };
    }
    
    const adminDb = mongoose.connection.db.admin();
    try {
      const dbInfo = await adminDb.ping();
      if (!dbInfo.ok) {
        errors.push('❌ Database ping failed - database may not be accessible');
      } else {
        successful.push('✅ Database is accessible');
        console.log('   ✅ Database Accessible');
      }
    } catch (pingError) {
      errors.push(`❌ Failed to ping database: ${pingError.message}`);
    }

    // Step 3: Verify required collections
    console.log('\n3️⃣  Verifying required collections...');
    const requiredCollections = ['sessions', 'drivers', 'vehicles', 'telemetry'];
    const existingCollections = await mongoose.connection.db.listCollections().toArray();
    const existingNames = existingCollections.map((c) => c.name);

    for (const collectionName of requiredCollections) {
      if (!existingNames.includes(collectionName)) {
        warnings.push(`⚠️  Collection "${collectionName}" does not exist yet (will be created on first write)`);
        console.log(`   ⚠️  ${collectionName} - Will be created on first write`);
      } else {
        successful.push(`✅ Collection "${collectionName}" exists`);
        console.log(`   ✅ ${collectionName} - Exists`);
      }
    }

    // Step 4: Test write permissions on each collection
    console.log('\n4️⃣  Testing write permissions...');
    const collections = [
      { name: 'sessions', model: Session },
      { name: 'drivers', model: Driver },
      { name: 'vehicles', model: Vehicle },
      { name: 'telemetry', model: Telemetry },
    ];

    for (const { name, model } of collections) {
      try {
        // Create a test document
        const testDoc = await model.collection.insertOne({
          __test_write__: true,
          timestamp: new Date(),
        });

        // Clean up - delete test document
        await model.collection.deleteOne({ _id: testDoc.insertedId });

        successful.push(`✅ Write test successful for "${name}"`);
        console.log(`   ✅ ${name} - Write OK`);
      } catch (writeError) {
        errors.push(
          `❌ Cannot write to collection "${name}": ${writeError.message}`
        );
        console.log(`   ❌ ${writeError.message}`);
      }
    }

    // Step 5: Verify indexes exist
    console.log('\n5️⃣  Verifying required indexes...');
    try {
      const sessionIndexes = await Session.collection.getIndexes();
      if (Object.keys(sessionIndexes).length > 1) {
        successful.push('✅ Session collection indexes are set up');
        console.log('   ✅ Session indexes OK');
      }

      const telemetryIndexes = await Telemetry.collection.getIndexes();
      // Should have at least sessionId index and TTL index
      const hasTTL = Object.values(telemetryIndexes).some((idx) => idx.expireAfterSeconds);
      if (hasTTL) {
        successful.push('✅ Telemetry TTL index is configured');
        console.log('   ✅ Telemetry TTL index OK');
      } else {
        warnings.push('⚠️  Telemetry TTL index not found (auto-cleanup disabled)');
        console.log('   ⚠️  No TTL index found');
      }
    } catch (indexError) {
      warnings.push(`⚠️  Could not verify indexes: ${indexError.message}`);
    }

    // Step 6: Check database storage stats
    console.log('\n6️⃣  Checking database storage...');
    try {
      const dbStats = await mongoose.connection.db.stats();
      const storageGB = dbStats.storageSize / (1024 * 1024 * 1024);
      const dataGB = dbStats.dataSize / (1024 * 1024 * 1024);

      console.log(`   Storage Used: ${storageGB.toFixed(2)} GB`);
      console.log(`   Data Size: ${dataGB.toFixed(2)} GB`);

      successful.push(`✅ Database storage stats: ${storageGB.toFixed(2)} GB used`);

      // Warn if storage getting full (assuming 5GB limit for development)
      if (storageGB > 4) {
        warnings.push(
          `⚠️  Database storage is ${storageGB.toFixed(2)} GB (limit ~5GB). Consider cleanup.`
        );
      }
    } catch (statsError) {
      warnings.push(`⚠️  Could not retrieve storage stats: ${statsError.message}`);
    }
  } catch (error) {
    errors.push(`❌ Unexpected validation error: ${error.message}`);
    console.error('   ❌ Unexpected error:', error.message);
  }

  // Summary
  console.log('\n' + '='.repeat(60));
  console.log('📊 VALIDATION SUMMARY');
  console.log('='.repeat(60));
  console.log(`✅ Successful: ${successful.length}`);
  console.log(`⚠️  Warnings: ${warnings.length}`);
  console.log(`❌ Errors: ${errors.length}`);

  if (successful.length > 0) {
    console.log('\n✅ Successful Checks:');
    successful.forEach((msg) => console.log(`   ${msg}`));
  }

  if (warnings.length > 0) {
    console.log('\n⚠️  Warnings:');
    warnings.forEach((msg) => console.log(`   ${msg}`));
  }

  if (errors.length > 0) {
    console.log('\n❌ Errors:');
    errors.forEach((msg) => console.log(`   ${msg}`));
  }

  console.log('='.repeat(60) + '\n');

  return {
    success: errors.length === 0,
    errors,
    warnings,
    successful,
    isReadyForSessions: errors.length === 0, // Strict: no sessions without all checks passing
  };
}

/**
 * Verify database is ready before allowing session operations
 * Called before session creation
 * @returns {Promise<boolean>} True if database is ready and valid
 */
export async function isDatabaseReadyForSessions() {
  try {
    // Quick check: can we write?
    const testCollection = mongoose.connection.db.collection('_db_health_check');
    const testId = `health_check_${Date.now()}`;

    // Try atomic operation
    const result = await testCollection.updateOne(
      { _id: testId },
      { $set: { timestamp: new Date() } },
      { upsert: true }
    );

    // Clean up
    await testCollection.deleteOne({ _id: testId });

    return result.acknowledged === true;
  } catch (error) {
    console.error('❌ Database health check failed:', error.message);
    return false;
  }
}

/**
 * Get database diagnostic information
 * Useful for monitoring and troubleshooting
 * @returns {Promise<Object>} Diagnostic info
 */
export async function getDatabaseDiagnostics() {
  const diagnostics = {
    connectionStatus: 'unknown',
    dbName: 'unknown',
    collectionsCount: 0,
    storageUsed: 0,
    dataSize: 0,
    indexCount: 0,
    errors: [],
  };

  try {
    // Connection status
    diagnostics.connectionStatus =
      mongoose.connection.readyState === 1 ? 'connected' : 'disconnected';

    // Database name
    diagnostics.dbName = mongoose.connection.db.getName();

    // Collection count
    const collections = await mongoose.connection.db.listCollections().toArray();
    diagnostics.collectionsCount = collections.length;

    // Storage stats
    const dbStats = await mongoose.connection.db.stats();
    diagnostics.storageUsed = dbStats.storageSize;
    diagnostics.dataSize = dbStats.dataSize;
    diagnostics.indexCount = dbStats.indexes;
  } catch (error) {
    diagnostics.errors.push(error.message);
  }

  return diagnostics;
}

File: web-app/ai-service/app/database_validator.py
"""
Database Validation Utility for AI-Service
Ensures MongoDB is ready and collections are writable before allowing sessions
Prevents silent failures from misconfigured collections or full database
"""

import asyncio
from motor.motor_asyncio import AsyncIOMotorClient, AsyncIOMotorDatabase
from app.utils import logger
from app.config import settings


class DatabaseValidator:
    """Validates database setup and writability"""

    def __init__(self, db_client: AsyncIOMotorClient = None, db: AsyncIOMotorDatabase = None):
        """
        Initialize validator with database connection
        
        Args:
            db_client: Motor AsyncIO MongoDB client
            db: Motor AsyncIO database instance
        """
        self.client = db_client
        self.db = db
        self.validation_result = None

    async def validate_database_setup(self) -> dict:
        """
        Comprehensive database validation
        Checks connection, collections, writability, and storage
        
        Returns:
            dict: Validation result with status, errors, warnings
        """
        logger.info("🔍 Starting comprehensive database validation...\n")
        
        errors = []
        warnings = []
        successful = []

        try:
            # Step 1: Verify connection
            logger.info("1️⃣  Testing MongoDB connection...")
            if self.db is None:
                errors.append("❌ MongoDB database instance not initialized")
                logger.error("   ❌ Database not initialized")
                return {
                    "success": False,
                    "errors": errors,
                    "warnings": warnings,
                    "successful": successful,
                    "is_ready_for_sessions": False,
                }

            try:
                # Ping the database
                result = await self.db.command("ping")
                if result.get("ok") == 1:
                    successful.append("✅ Connected to MongoDB")
                    logger.info("   ✅ Connected")
                else:
                    errors.append("❌ Database ping failed")
                    logger.error("   ❌ Ping failed")
            except Exception as ping_error:
                errors.append(f"❌ Failed to ping database: {str(ping_error)}")
                logger.error(f"   ❌ Ping error: {ping_error}")

            # Step 2: Verify database and collection name configuration
            logger.info("\n2️⃣  Verifying database and collection configuration...")
            if not settings.COLLECTION_NAME:
                errors.append("❌ COLLECTION_NAME not configured in .env")
                logger.error("   ❌ COLLECTION_NAME missing from config")
            else:
                logger.info(f"   Collection: {settings.COLLECTION_NAME}")
                successful.append(f"✅ Collection name configured: {settings.COLLECTION_NAME}")

            # Step 3: List existing collections
            logger.info("\n3️⃣  Checking for required collections...")
            try:
                collections = await self.db.list_collection_names()
                logger.info(f"   Found {len(collections)} existing collection(s)")
                
                if settings.COLLECTION_NAME not in collections:
                    warnings.append(
                        f"⚠️  Collection '{settings.COLLECTION_NAME}' does not exist yet "
                        f"(will be created upon first event insert)"
                    )
                    logger.info(f"   ⚠️  {settings.COLLECTION_NAME} - Will be created on first write")
                else:
                    successful.append(f"✅ Collection '{settings.COLLECTION_NAME}' exists")
                    logger.info(f"   ✅ {settings.COLLECTION_NAME} - Exists")
            except Exception as list_error:
                errors.append(f"❌ Could not list collections: {str(list_error)}")
                logger.error(f"   ❌ Error listing collections: {list_error}")

            # Step 4: Test write permissions
            logger.info("\n4️⃣  Testing write permissions...")
            try:
                collection = self.db[settings.COLLECTION_NAME]
                
                # Insert test document
                test_doc = {
                    "__test_write__": True,
                    "timestamp": asyncio.get_event_loop().time(),
                }
                result = await collection.insert_one(test_doc)
                test_id = result.inserted_id
                logger.info(f"   Inserted test document: {test_id}")
                
                # Delete test document
                await collection.delete_one({"_id": test_id})
                logger.info("   Deleted test document")
                
                successful.append(f"✅ Write test successful for '{settings.COLLECTION_NAME}'")
                logger.info(f"   ✅ {settings.COLLECTION_NAME} - Write OK")
            except Exception as write_error:
                errors.append(
                    f"❌ Cannot write to collection '{settings.COLLECTION_NAME}': "
                    f"{str(write_error)}"
                )
                logger.error(f"   ❌ Write failed: {write_error}")

            # Step 5: Check for indexes
            logger.info("\n5️⃣  Checking collection indexes...")
            try:
                collection = self.db[settings.COLLECTION_NAME]
                indexes = await collection.list_indexes().to_list(None)
                
                if len(indexes) > 1:  # More than just _id
                    successful.append(f"✅ Collection has {len(indexes)} index(es)")
                    logger.info(f"   ✅ Indexes found: {len(indexes)}")
                else:
                    warnings.append("⚠️  Collection has no custom indexes (only default _id)")
                    logger.info("   ⚠️  No custom indexes")
            except Exception as index_error:
                warnings.append(f"⚠️  Could not check indexes: {str(index_error)}")
                logger.warning(f"   ⚠️  Index check error: {index_error}")

            # Step 6: Check database storage stats
            logger.info("\n6️⃣  Checking database storage...")
            try:
                db_stats = await self.db.command("dbstats")
                
                storage_mb = db_stats.get("storageSize", 0) / (1024 * 1024)
                data_mb = db_stats.get("dataSize", 0) / (1024 * 1024)
                collections_count = db_stats.get("collections", 0)
                
                logger.info(f"   Storage: {storage_mb:.2f} MB")
                logger.info(f"   Data: {data_mb:.2f} MB")
                logger.info(f"   Collections: {collections_count}")
                
                successful.append(f"✅ Storage: {storage_mb:.2f} MB used")
                
                # Warn if approaching typical limits
                if storage_mb > 4000:  # 4GB warning
                    warnings.append(
                        f"⚠️  Database storage is {storage_mb:.2f} MB. "
                        f"Consider cleanup if approaching limits."
                    )
                    logger.warning(f"   ⚠️  High storage usage: {storage_mb:.2f} MB")
                    
            except Exception as stats_error:
                warnings.append(f"⚠️  Could not retrieve storage stats: {str(stats_error)}")
                logger.warning(f"   ⚠️  Stats error: {stats_error}")

        except Exception as error:
            errors.append(f"❌ Unexpected validation error: {str(error)}")
            logger.error(f"   ❌ Unexpected error: {error}")

        # Summary
        logger.info("\n" + "=" * 60)
        logger.info("📊 VALIDATION SUMMARY")
        logger.info("=" * 60)
        logger.info(f"✅ Successful: {len(successful)}")
        logger.info(f"⚠️  Warnings: {len(warnings)}")
        logger.info(f"❌ Errors: {len(errors)}")

        if successful:
            logger.info("\n✅ Successful Checks:")
            for msg in successful:
                logger.info(f"   {msg}")

        if warnings:
            logger.info("\n⚠️  Warnings:")
            for msg in warnings:
                logger.info(f"   {msg}")

        if errors:
            logger.info("\n❌ Errors:")
            for msg in errors:
                logger.info(f"   {msg}")

        logger.info("=" * 60 + "\n")

        self.validation_result = {
            "success": len(errors) == 0,
            "errors": errors,
            "warnings": warnings,
            "successful": successful,
            "is_ready_for_sessions": len(errors) == 0,  # Strict: no sessions without passing all checks
        }

        return self.validation_result

    async def is_database_ready_for_sessions(self) -> bool:
        """
        Quick check: Is database ready to accept sessions?
        Returns True only if database passed validation
        
        Returns:
            bool: True if database is ready and writable
        """
        if self.db is None:
            logger.error("❌ Database not initialized")
            return False

        try:
            # Quick health check: Can we write?
            health_collection = self.db["_health_check"]
            test_id = f"health_check_{asyncio.get_event_loop().time()}"

            result = await health_collection.update_one(
                {"_id": test_id},
                {"$set": {"timestamp": asyncio.get_event_loop().time()}},
                upsert=True
            )

            # Clean up
            await health_collection.delete_one({"_id": test_id})

            return result.acknowledged is True
        except Exception as error:
            logger.error(f"❌ Database health check failed: {error}")
            return False

    async def get_database_diagnostics(self) -> dict:
        """
        Get diagnostic information about database
        Useful for monitoring and troubleshooting
        
        Returns:
            dict: Diagnostic information
        """
        diagnostics = {
            "db_name": None,
            "collections_count": 0,
            "storage_mb": 0,
            "data_mb": 0,
            "errors": [],
        }

        try:
            if self.db is None:
                diagnostics["errors"].append("Database not initialized")
                return diagnostics

            # Database name
            diagnostics["db_name"] = self.db.name

            # Collection count
            try:
                collections = await self.db.list_collection_names()
                diagnostics["collections_count"] = len(collections)
            except Exception as e:
                diagnostics["errors"].append(f"Could not list collections: {str(e)}")

            # Storage stats
            try:
                db_stats = await self.db.command("dbstats")
                diagnostics["storage_mb"] = db_stats.get("storageSize", 0) / (1024 * 1024)
                diagnostics["data_mb"] = db_stats.get("dataSize", 0) / (1024 * 1024)
            except Exception as e:
                diagnostics["errors"].append(f"Could not get storage stats: {str(e)}")

        except Exception as error:
            diagnostics["errors"].append(f"Unexpected diagnostics error: {str(error)}")
            logger.error(f"Diagnostics error: {error}")

        return diagnostics


# Convenience functions for use throughout AI-service

async def validate_db_on_startup(db_client: AsyncIOMotorClient, db: AsyncIOMotorDatabase) -> bool:
    """
    Validate database during application startup
    Fails startup if database validation fails
    
    Args:
        db_client: Motor AsyncIO client
        db: Motor AsyncIO database
        
    Returns:
        bool: True if validation passed
    """
    validator = DatabaseValidator(db_client, db)
    result = await validator.validate_database_setup()
    
    if not result["is_ready_for_sessions"]:
        logger.error("🛑 DATABASE VALIDATION FAILED")
        logger.error("❌ Application will not start without a valid database")
        raise RuntimeError(
            "Database validation failed. Please check MongoDB connection and configuration."
        )

    logger.info("✅ DATABASE VALIDATION PASSED")
    logger.info("✅ Application ready to accept sessions")
    return True